### PR TITLE
HW-01. Autoregressive models

### PR DESCRIPTION
**Pull request by Javanochka**

### CODE REVIEW

**Comment by Rauf-Kurbanov** [comment]
Задача 1:
* Можно было использовать `torch.nn.Softmax`, а лучше `torch.nn.LogSoftmax`
* Хорошая практика фиксировать random seed
* NLL не в битах

А в целом все хорошо - зачтено

**Comment by Rauf-Kurbanov** [comment]
Задача 2.1:
* Много дублирования кода
* `F.softmax` is deprecated 
* Лучше вместо `requires_grad=True` оборачивать в `torch.nn.Parameter`
* Если бы обернула решение задачи 1 в `nn.Module`, могла бы переиспользовать и не дублировать код.
* В `ModelSimple` полную вероятность можно было посчитать и векторно

**Comment by Rauf-Kurbanov** [comment]
Задача 2.2

Скорее всего косяк в масках: тебе на самом деле не нужно так много слоев ни в 2.1 ни в 2.2. Попробуй просто нарисовать маску на бумажке - там всё просто

**Comment by Rauf-Kurbanov** [comment]
Задача 3
Жду решения

**Comment by Javanochka** [comment]
@Rauf-Kurbanov Во второй части, как оказалось, была проблема не в масках, а в обучении (иногда log 0, иногда не обнулялся градиент)

Доделала. Вроде бы сделала часть три.

**Review by Rauf-Kurbanov — APPROVED** [review]
all approved
